File: app/src/main/java/net/aquadx/aquacard/data/AquaProfileRepository.kt
package net.aquadx.aquacard.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.supervisorScope
import kotlinx.coroutines.withContext

/**
 * Оркестрация загрузки профиля maimai (docs/profile-ui-plan.md §3.3).
 * Все секции грузятся параллельно; try/catch ВНУТРИ каждого async, поэтому отказ
 * одной секции не отменяет соседей (supervisorScope) и пишется в errors.
 */
class AquaProfileRepository(
    private val service: AquaProfileService,
    private val metaCache: MutableMap<String, Map<Int, MusicMeta>> = sharedMetaCache
) {

    suspend fun load(baseUrl: String, username: String): ProfileBundle =
        withContext(Dispatchers.IO) {
            val errors = mutableListOf<String>()
            supervisorScope {
                val summaryD = async { section(errors, "summary") { service.summary(username) } }
                val detailD = async { section(errors, "detail") { service.detail(username) } }
                val ratingD = async { section(errors, "rating") { service.rating(username) } }
                val recentD = async { section(errors, "recent") { service.recent(username) } }
                val trendD = async { section(errors, "trend") { service.trend(username) } }
                val metaD = async { section(errors, "meta") { loadMeta(baseUrl) } }

                val rating = ratingD.await()
                ProfileBundle(
                    summary = summaryD.await(),
                    detail = detailD.await(),
                    best = mapBest(rating?.best35 ?: emptyList()),
                    bestSecondary = mapBest(rating?.best15 ?: emptyList()),
                    scores = mapScores(rating?.musicList ?: emptyList()),
                    recent = mapRecent(recentD.await() ?: emptyList()),
                    trend = trendD.await() ?: emptyList(),
                    meta = metaD.await() ?: emptyMap(),
                    errors = errors
                )
            }
        }

    private suspend fun loadMeta(baseUrl: String): Map<Int, MusicMeta> {
        metaCache[GAME]?.let { return it }
        val raw = service.allMusic(allMusicUrl(baseUrl))
        val parsed = raw.mapNotNull { (key, value) -> key.toIntOrNull()?.let { it to value } }.toMap()
        metaCache[GAME] = parsed
        return parsed
    }

    /** Inline → suspend-вызовы внутри block разрешены; ошибка секции изолируется в null. */
    private inline fun <T> section(errors: MutableList<String>, name: String, block: () -> T): T? =
        try {
            block()
        } catch (e: Exception) {
            synchronized(errors) { errors.add("$name: ${e.message ?: e::class.simpleName ?: "error"}") }
            null
        }

    private fun mapBest(tuples: List<List<String>>): List<BestEntry> =
        tuples.mapNotNull { ScoreFormat.parseBestTuple(it) }

    private fun mapScores(list: List<MusicScoreDto>): List<ProfileScore> =
        list.map { dto ->
            ProfileScore(
                musicId = dto.musicId,
                level = dto.level ?: 0,
                achievement = dto.achievement,
                deluxscore = dto.deluxscoreMax,
                comboStatus = dto.comboStatus,
                syncStatus = dto.syncStatus,
                scoreRank = dto.scoreRank
            )
        }.sortedByDescending { it.achievement ?: 0 }

    // /recent отдаёт партии по возрастанию (старые первыми) — сортируем по дате УБЫВАНИЕМ,
    // чтобы «последние» были действительно последними. Дата нормализуется (срез хвостового ".0").
    private fun mapRecent(list: List<RecentPlayDto>): List<RecentEntry> =
        list.sortedByDescending { (it.userPlayDate ?: it.playDate ?: "").substringBefore('.') }
            .map { dto ->
                RecentEntry(
                    musicId = dto.musicId,
                    level = dto.level ?: 0,
                    playDate = dto.userPlayDate ?: dto.playDate,
                    achievement = dto.achievement,
                    rank = dto.scoreRank,
                    comboStatus = dto.comboStatus,
                    isClear = dto.isClear,
                    syncStatus = dto.syncStatus,
                    deluxscore = dto.deluxscore,
                    beforeRating = dto.beforeRating,
                    afterRating = dto.afterRating,
                    placeName = dto.placeName,
                    maxCombo = dto.maxCombo,
                    totalCombo = dto.totalCombo,
                    fastCount = dto.fastCount,
                    lateCount = dto.lateCount,
                    isFullCombo = dto.isFullCombo,
                    isAllPerfect = dto.isAllPerfect,
                    trackNo = dto.trackNo,
                    judges = judgeBreakdown(dto),
                    notes = noteBreakdown(dto)
                )
            }

    private fun judgeBreakdown(d: RecentPlayDto): JudgeBreakdown? {
        val crit = sum(d.tapCriticalPerfect, d.holdCriticalPerfect, d.slideCriticalPerfect, d.touchCriticalPerfect, d.breakCriticalPerfect)
        val perfect = sum(d.tapPerfect, d.holdPerfect, d.slidePerfect, d.touchPerfect, d.breakPerfect)
        val great = sum(d.tapGreat, d.holdGreat, d.slideGreat, d.touchGreat, d.breakGreat)
        val good = sum(d.tapGood, d.holdGood, d.slideGood, d.touchGood, d.breakGood)
        val miss = sum(d.tapMiss, d.holdMiss, d.slideMiss, d.touchMiss, d.breakMiss)
        return if (crit + perfect + great + good + miss == 0) null
        else JudgeBreakdown(crit, perfect, great, good, miss)
    }

    private fun noteBreakdown(d: RecentPlayDto): NoteBreakdown? {
        val tap = sum(d.tapCriticalPerfect, d.tapPerfect, d.tapGreat, d.tapGood, d.tapMiss)
        val hold = sum(d.holdCriticalPerfect, d.holdPerfect, d.holdGreat, d.holdGood, d.holdMiss)
        val slide = sum(d.slideCriticalPerfect, d.slidePerfect, d.slideGreat, d.slideGood, d.slideMiss)
        val touch = sum(d.touchCriticalPerfect, d.touchPerfect, d.touchGreat, d.touchGood, d.touchMiss)
        val brk = sum(d.breakCriticalPerfect, d.breakPerfect, d.breakGreat, d.breakGood, d.breakMiss)
        return if (tap + hold + slide + touch + brk == 0) null
        else NoteBreakdown(tap, hold, slide, touch, brk)
    }

    private fun sum(vararg v: Int?): Int = v.sumOf { it ?: 0 }

    companion object {
        private const val GAME = "mai2"
        val sharedMetaCache = mutableMapOf<String, Map<Int, MusicMeta>>()

        /** Статик-хост ассетов из API-базы: убираем хвостовой /aqua, чтобы получить корень. */
        fun staticBase(baseUrl: String): String {
            val trimmed = baseUrl.trim().removeSuffix("/")
            return if (trimmed.endsWith("/aqua")) trimmed.removeSuffix("/aqua") else trimmed
        }

        fun allMusicUrl(baseUrl: String): String =
            "${staticBase(baseUrl)}/d/$GAME/00/all-music.json"
    }
}

File: iosApp/AquaCard/AquaProfileRepository.swift
import Foundation

struct AquaProfileRepository {
    let client: AquaProfileClient
    let cache: ProfileCacheStore

    func readCache(baseURL: String, username: String) async -> CachedProfile? {
        await cache.read(baseURL: baseURL, username: username)
    }

    func writeCache(baseURL: String, username: String, bundle: ProfileBundle) async {
        await cache.write(baseURL: baseURL, username: username, bundle: bundle)
    }

    func load(baseURL: String, username: String) async throws -> ProfileBundle {
        async let summaryResult = section("summary") { try await client.summary(baseURL: baseURL, username: username) }
        async let detailResult = section("detail") { try await client.detail(baseURL: baseURL, username: username) }
        async let ratingResult = section("rating") { try await client.rating(baseURL: baseURL, username: username) }
        async let recentResult = section("recent") { try await client.recent(baseURL: baseURL, username: username) }
        async let trendResult = section("trend") { try await client.trend(baseURL: baseURL, username: username) }
        async let metaResult = musicMetadata(baseURL: baseURL)

        try Task.checkCancellation()
        let summary = await summaryResult
        let detail = await detailResult
        let rating = await ratingResult
        let recent = await recentResult
        let trend = await trendResult
        let meta = await metaResult
        try Task.checkCancellation()

        var errors = [String]()
        errors.append(contentsOf: [summary.error, detail.error, rating.error, recent.error, trend.error, meta.error].compactMap { $0 })

        if summary.value == nil,
           detail.value == nil,
           rating.value == nil,
           recent.value == nil,
           trend.value == nil {
            throw AquaRepositoryError.noProfileSections(errors)
        }

        return ProfileBundle(
            summary: summary.value,
            detail: detail.value,
            best: mapBest(rating.value?.best35 ?? []),
            bestSecondary: mapBest(rating.value?.best15 ?? []),
            scores: mapScores(rating.value?.musicList ?? []),
            recent: mapRecent(recent.value ?? []),
            trend: trend.value ?? [],
            meta: meta.value ?? [:],
            errors: errors
        )
    }

    private func section<T>(_ name: String, block: () async throws -> T) async -> SectionResult<T> {
        do {
            try Task.checkCancellation()
            return SectionResult(value: try await block(), error: nil)
        } catch is CancellationError {
            return SectionResult(value: nil, error: nil)
        } catch {
            return SectionResult(value: nil, error: "\(name): \(error.localizedDescription)")
        }
    }

    private func musicMetadata(baseURL: String) async -> SectionResult<[Int: MusicMeta]> {
        let key = AquaAssets.staticBase(baseURL)
        if let cached = await MusicMetadataCache.shared.value(for: key) {
            return SectionResult(value: cached, error: nil)
        }
        do {
            try Task.checkCancellation()
            let loaded = try await client.allMusic(baseURL: baseURL)
            try Task.checkCancellation()
            await MusicMetadataCache.shared.store(loaded, for: key)
            return SectionResult(value: loaded, error: nil)
        } catch is CancellationError {
            return SectionResult(value: nil, error: nil)
        } catch {
            return SectionResult(value: nil, error: "meta: \(error.localizedDescription)")
        }
    }

    private func mapBest(_ tuples: [[String]]) -> [BestEntry] {
        tuples.compactMap { ScoreFormat.parseBestTuple($0) }
    }

    private func mapScores(_ list: [MusicScoreDto]) -> [ProfileScore] {
        list.map {
            ProfileScore(
                musicId: $0.musicId,
                level: $0.level ?? 0,
                achievement: $0.achievement,
                deluxscore: $0.deluxscoreMax,
                comboStatus: $0.comboStatus,
                syncStatus: $0.syncStatus,
                scoreRank: $0.scoreRank
            )
        }.sorted { ($0.achievement ?? 0) > ($1.achievement ?? 0) }
    }

    private func mapRecent(_ list: [RecentPlayDto]) -> [RecentEntry] {
        list.sorted {
            ($0.userPlayDate ?? $0.playDate ?? "").split(separator: ".").first ?? "" >
            ($1.userPlayDate ?? $1.playDate ?? "").split(separator: ".").first ?? ""
        }.map {
            RecentEntry(
                musicId: $0.musicId,
                level: $0.level ?? 0,
                playDate: $0.userPlayDate ?? $0.playDate,
                achievement: $0.achievement,
                rank: $0.scoreRank,
                comboStatus: $0.comboStatus,
                isClear: $0.isClear,
                syncStatus: $0.syncStatus,
                deluxscore: $0.deluxscore,
                beforeRating: $0.beforeRating,
                afterRating: $0.afterRating,
                placeName: $0.placeName,
                maxCombo: $0.maxCombo,
                totalCombo: $0.totalCombo,
                fastCount: $0.fastCount,
                lateCount: $0.lateCount,
                isFullCombo: $0.isFullCombo,
                isAllPerfect: $0.isAllPerfect,
                trackNo: $0.trackNo,
                judges: judgeBreakdown($0),
                notes: noteBreakdown($0)
            )
        }
    }

    private func judgeBreakdown(_ dto: RecentPlayDto) -> JudgeBreakdown? {
        let crit = sum(dto.tapCriticalPerfect, dto.holdCriticalPerfect, dto.slideCriticalPerfect, dto.touchCriticalPerfect, dto.breakCriticalPerfect)
        let perfect = sum(dto.tapPerfect, dto.holdPerfect, dto.slidePerfect, dto.touchPerfect, dto.breakPerfect)
        let great = sum(dto.tapGreat, dto.holdGreat, dto.slideGreat, dto.touchGreat, dto.breakGreat)
        let good = sum(dto.tapGood, dto.holdGood, dto.slideGood, dto.touchGood, dto.breakGood)
        let miss = sum(dto.tapMiss, dto.holdMiss, dto.slideMiss, dto.touchMiss, dto.breakMiss)
        return crit + perfect + great + good + miss == 0 ? nil : JudgeBreakdown(crit: crit, perfect: perfect, great: great, good: good, miss: miss)
    }

    private func noteBreakdown(_ dto: RecentPlayDto) -> NoteBreakdown? {
        let tap = sum(dto.tapCriticalPerfect, dto.tapPerfect, dto.tapGreat, dto.tapGood, dto.tapMiss)
        let hold = sum(dto.holdCriticalPerfect, dto.holdPerfect, dto.holdGreat, dto.holdGood, dto.holdMiss)
        let slide = sum(dto.slideCriticalPerfect, dto.slidePerfect, dto.slideGreat, dto.slideGood, dto.slideMiss)
        let touch = sum(dto.touchCriticalPerfect, dto.touchPerfect, dto.touchGreat, dto.touchGood, dto.touchMiss)
        let brk = sum(dto.breakCriticalPerfect, dto.breakPerfect, dto.breakGreat, dto.breakGood, dto.breakMiss)
        return tap + hold + slide + touch + brk == 0 ? nil : NoteBreakdown(tap: tap, hold: hold, slide: slide, touch: touch, brk: brk)
    }

    private func sum(_ values: Int?...) -> Int {
        values.reduce(0) { $0 + ($1 ?? 0) }
    }
}

private struct SectionResult<T> {
    var value: T?
    var error: String?
}

private actor MusicMetadataCache {
    static let shared = MusicMetadataCache()
    private let maxEntries = 3
    private var cachedByStaticBase: [String: [Int: MusicMeta]] = [:]
    private var recency: [String] = []

    func value(for key: String) -> [Int: MusicMeta]? {
        guard let value = cachedByStaticBase[key] else { return nil }
        markRecent(key)
        return value
    }

    func store(_ value: [Int: MusicMeta], for key: String) {
        cachedByStaticBase[key] = value
        markRecent(key)
        while recency.count > maxEntries, let oldest = recency.first {
            recency.removeFirst()
            cachedByStaticBase.removeValue(forKey: oldest)
        }
    }

    private func markRecent(_ key: String) {
        recency.removeAll { $0 == key }
        recency.append(key)
    }
}

enum AquaRepositoryError: LocalizedError {
    case noProfileSections([String])

    var errorDescription: String? {
        switch self {
        case .noProfileSections(let errors):
            return errors.isEmpty ? "No profile data was returned" : errors.joined(separator: "; ")
        }
    }
}
